src/core/derivation.ts

```diff
@@ -136,37 +148,37 @@ function bindDependencies(derivation: IDerivation, prevObserving: IObservable[])
 
 	// This process is optimized by making sure deps are always left 'clean', with value 0, so that they don't need to be reset at the start of this process
 	// after that, all prevObserving items are marked with -1 directly, instead of 0 and doing -- after that
-	// further the +1 and addObserver can be done in one go.
-	for (let i = 0; i < prevLength; i++)
-		prevObserving[i].diffValue = -1;
-
-	for (let i = 0; i < newLength; i++) {
+	for (let i = 0; i < observingLength; i++) {
 		const dep = observing[i];
-		// there is no guarantee that the observing collection is unique, so especially in the first run of the derivation
-		// this check might succeed too often (namely, for each double dep). That is no problem because addObserve is backed by a set
-		// In subsequent runs of the derivation, double entries are a lot less likely to happen, because then the used derivations are hot and executed
-		// _before_ this derivation, meaning that the lastAccessedDerivation optimization will most probably have skip all the doubles already.
-		// see also the "unoptimizable subscriptions are diffed correctly" test
-		if ((++dep.diffValue) > 0) {
-			dep.diffValue = 0; // this also short circuits add if a dep is multiple times in the observing list
-			addObserver(dep, derivation);
+		if ((++dep.diffValue) === 1) {
+			uniqueObserving.push(dep);
 		}
 	}
 
+	// further the -1 and removeObserver can be done in one go.
 	for (let i = 0; i < prevLength; i++) {
 		const dep = prevObserving[i];
-		if (dep.diffValue < 0) {
+		if (--prevObserving[i].diffValue === -1) {
 			dep.diffValue = 0; // this also short circuits add if a dep is multiple times in the observing list
 			removeObserver(dep, derivation);
 		}
 	}
+
+	for (let i = 0; i < uniqueObserving.length; i++) {
+		const dep = uniqueObserving[i];
+		if (dep.diffValue > 0) {
+			dep.diffValue = 0; // this also short circuits add if a dep is multiple times in the observing list
+			addObserver(dep, derivation);
+		}
+	}
 }
 
 export function clearObserving(derivation: IDerivation) {
 	const obs = derivation.observing;
 	const l = obs.length;
 	for (let i = 0; i < l; i++)
 		removeObserver(obs[i], derivation);
+	derivation.dependenciesState = -1;
```

Comment:
might be clearer to have changeDependencieState(-1) and clear the observers there?